coderd/wsbuilder/wsbuilder.go

```diff
@@ -0,0 +1,741 @@
+}
+
+// Build computes and inserts a new workspace build into the database.  If authFunc is provided, it also performs
+// authorization preflight checks.
+func (b *Builder) Build(
+	ctx context.Context,
+	store database.Store,
+	authFunc func(action rbac.Action, object rbac.Objecter) bool,
+) (
+	*database.WorkspaceBuild, *database.ProvisionerJob, error,
+) {
+	b.ctx = ctx
+
+	// Run the build in a transaction with RepeatableRead isolation, and retries.
+	// RepeatableRead isolation ensures that we get a consistent view of the database while
+	// computing the new build.  This simplifies the logic so that we do not need to worry if
+	// later reads are consistent with earlier ones.
+	var err error
+	for retries := 0; retries < 5; retries++ {
+		var workspaceBuild *database.WorkspaceBuild
+		var provisionerJob *database.ProvisionerJob
+		err := store.InTx(func(store database.Store) error {
+			b.store = store
+			workspaceBuild, provisionerJob, err = b.buildTx(authFunc)
+			return err
+		}, &sql.TxOptions{Isolation: sql.LevelRepeatableRead})
+		var pqe *pq.Error
+		if xerrors.As(err, &pqe) {
+			if pqe.Code == "40001" {
+				// serialization error, retry
+				continue
+			}
+		}
+		if err != nil {
+			// Other (hard) error
+			return nil, nil, err
+		}
+		return workspaceBuild, provisionerJob, nil
+	}
+	return nil, nil, xerrors.Errorf("too many errors; last error: %w", err)
+}
+
+// buildTx contains the business logic of computing a new build.  Attributes of the new database objects are computed
+// in a functional style, rather than imperative, to emphasize the logic of how they are defined.  A simple cache
+// of database-fetched objects is stored on the struct to ensure we only fetch things once, even if they are used in
+// the calculation of multiple attributes.
+//
+// In order to utilize this cache, the functions that compute build attributes use a pointer receiver type.
+func (b *Builder) buildTx(authFunc func(action rbac.Action, object rbac.Objecter) bool) (
+	*database.WorkspaceBuild, *database.ProvisionerJob, error,
+) {
+	if authFunc != nil {
+		err := b.authorize(authFunc)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	err := b.checkTemplateVersionMatchesTemplate()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = b.checkTemplateJobStatus()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = b.checkRunningBuild()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	template, err := b.getTemplate()
+	if err != nil {
+		return nil, nil, BuildError{http.StatusInternalServerError, "failed to fetch template", err}
+	}
+
+	templateVersionJob, err := b.getTemplateVersionJob()
+	if err != nil {
+		return nil, nil, BuildError{
+			http.StatusInternalServerError, "failed to fetch template version job", err,
+		}
+	}
+
+	legacyParameters, err := b.getLastParameterValues()
+	if err != nil {
+		return nil, nil, BuildError{
+			http.StatusInternalServerError,
+			"failed to fetch previous legacy parameters.",
+			err,
+		}
+	}
```

Comment:
Seems like all of these can happen in parallel and use an errgroup.

Reply:
That adds a fair amount of complexity.

The method calls read & modify an internal cache of db objects, which would have to have a mutex protecting it if it gets accessed from multiple goroutines.  Holding that lock while querying the database would bring us back serial requests, but not holding the lock opens a big can of worms where you have to track requests in flight, lest we query the database multiple times for the same thing.  So, the database calls get partially serialized as a natural consequence, meaning the level of parallelism we can achieve isn't that great.

I say not worth it!